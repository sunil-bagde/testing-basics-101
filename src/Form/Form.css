form {
    display: flex;
    flex-direction: column;
    width: 80%;
    margin: 50px auto;
    padding: 35px;
}
input {
    margin: 10px 0;
    border: solid 1px #005fcc;
    border-radius: 0px;
    padding: 7px;
}
button {
    background: skyblue;
    border: none;
    border-radius: 30px;
    width: 300px;
    color: black;
    margin: auto;
    text-transform: uppercase;
    letter-spacing: 4px;
    font-size: 20px;
    padding: 10px 0;
    margin-top: 25px;
    cursor: pointer;
}
button:hover {
    background: #4fbeeb;
}
h2 {
    text-transform: uppercase;
    font-weight: 100;
    letter-spacing: 3px;
}
.agree {
    display: inline-block;
    margin-left: 10px;
}
.alert {
    padding: 0;
    margin: 0;
    color: red;
}
