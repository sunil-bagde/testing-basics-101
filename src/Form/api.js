export default {
    addUser : (name, email, number) => new Promise((res,rej) => res({name,email,number}))
}