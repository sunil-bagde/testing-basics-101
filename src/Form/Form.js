import React, { Component } from "react";
import "./Form.css";
import api from "./api";

export default class Form extends Component {
    state = {
        name: "",
        email: "",
        number: "",
        optIn: true,
        erros: {
            _name: "",
            _email: "",
            _number: "",
            _optIn: "",
        },
    };
    handleChange = (str) => (e) => {
        this.setState({ [str]: e.currentTarget.value });
    };
    handleSubmit = async (e) => {
        e.preventDefault();

        const name = this.state.name;
        const email = this.state.email;
        const number = this.state.number;
        const optIn = this.state.optIn;
        const erros = { ...this.state };
        if (name.length === 0) {
            erros["_name"] = "Name is required";
        }
        if (email.length === 0) {
            erros["_email"] = "Email is required";
        }
        if (number.length === 0) {
            erros["_number"] = "Number is required";
        }
        if (optIn === false) {
            erros["_optIn"] = "optIn is required";
        }
   
        let hasAnyErros = Object.keys(erros).every((e) => e.length === 0);

        if (!hasAnyErros) {
            this.setState({
                erros,
            });
        }
 

        const user = await api.addUser(
            this.state.name,
            this.state.email,
            this.state.number
        );
        if (user) {
            this.setState({
                name: "",
                email: "",
                number: "",
                optIn: false,
            });
        }
    };
    handlePromotionClick = (e) => {
        this.setState((prevState) => ({ optIn: !prevState.optIn }));
    };
    render() {
        const { erros } = this.state;
        //   const { name, number, email, optIn } = erros;

        return (
            <form data-testid="addUserForm" onSubmit={this.handleSubmit}>
                <input
                    data-testid="name"
                    type="text"
                    onChange={this.handleChange("name")}
                    placeholder="Name"
                    value={this.state.name}
                />
                <p className="alert">{erros._name}</p>

                <input
                    data-testid="email"
                    type="text"
                    onChange={this.handleChange("email")}
                    placeholder="Email"
                    value={this.state.email}
                />
                <p className="alert">{erros._email}</p>
                <input
                    data-testid="number"
                    type="text"
                    onChange={this.handleChange("number")}
                    placeholder="Mobile"
                    value={this.state.number}
                />
                <p className="alert">{erros._number}</p>
                <div>
                    <input
                        data-testid="checked"
                        type="checkbox"
                        checked={this.state.optIn}
                        onChange={() => {}}
                        onClick={this.handlePromotionClick}
                    />
                    <p data-testid="promotionsP" className="agree">
                        Accept terms
                    </p>
                    <p className="alert">{erros._optIn}</p>
                </div>
                <button type="submit" data-testid="submitButton">
                    Submit
                </button>
            </form>
        );
    }
}
