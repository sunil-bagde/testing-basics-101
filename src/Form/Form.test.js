import React from "react";
import Form from "./Form";
import { shallow, configure } from "enzyme";
import toJson from "enzyme-to-json";
import api from "./api";

function updateInput(wrapper, instance, newValue) {
    const input = wrapper.find(instance);
    input.simulate("change", {
        currentTarget: { value: newValue },
    });
    return wrapper.find(instance);
}
function getByTestId(name) {
    return `[data-testid="${name}"]`;
}
describe("<Form />", () => {
    test("Terms should default checked ", () => {
        const wrapper = shallow(<Form />);
        const terms = wrapper.find(getByTestId("checked"));
        const checker = terms.props().checked;
        expect(checker).toBe(true);
    });

    test("user should to fill form", () => {
        const wrapper = shallow(<Form />);

        const nameInput = updateInput(wrapper, getByTestId("name"), "admin");

        expect(nameInput.props().value).toBe("admin");

        const emailInput = updateInput(
            wrapper,
            getByTestId("email"),
            "admin@gmail.com"
        );
        const numberInput = updateInput(
            wrapper,
            getByTestId("number"),
            "8655062213"
        );
        wrapper.find(getByTestId("checked")).simulate("click");

        expect(nameInput.props().value).toBe("admin");
        expect(emailInput.props().value).toBe("admin@gmail.com");
        expect(numberInput.props().value).toBe("8655062213");
        expect(wrapper.find(getByTestId("checked")).props().checked).toBe(
            false
        );
    });
    test("matches saved snapshot", async () => {
        const wrapper = shallow(<Form />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

     test("submits fails when validation errors ", () => {
        let response = null;
        jest.spyOn(api, "addUser").mockImplementation(
            () => (response = Promise.resolve({ data: "New User Added" }))
        );
        const wrapper = shallow(<Form />);
        updateInput(wrapper, getByTestId("name"), "");
        updateInput(wrapper, getByTestId("email"), "");
        updateInput(wrapper, getByTestId("number"), "");
        updateInput(wrapper, getByTestId("checked"), false);
                wrapper
            .find(getByTestId("addUserForm"))
            .simulate("submit", { preventDefault: () => {} });
          //  wrapper.find(".alert").toMatchSnapshot();

           expect(toJson(wrapper.find(".alert"))).toMatchSnapshot();
    });

    test("submits the form", () => {
        let response = null;
        jest.spyOn(api, "addUser").mockImplementation(() =>
          response =   Promise.resolve({ data: "New User Added" })
        );

        const wrapper = shallow(<Form />);
        updateInput(wrapper, getByTestId("name"), "admin");
        updateInput(wrapper, getByTestId("email"), "test@gmail.com");
        updateInput(wrapper, getByTestId("number"), "8655091122");
        wrapper
            .find( getByTestId("addUserForm"))
            .simulate("submit", { preventDefault: () => {} });
        expect(api.addUser).toHaveBeenCalledWith(
            "admin",
            "test@gmail.com",
            "8655091122"
        );
        response.then(r => {
            expect(r.data).toMatchSnapshot()
        })
    });
});

 


 
