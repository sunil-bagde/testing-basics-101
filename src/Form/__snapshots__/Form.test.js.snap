// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> matches saved snapshot 1`] = `
<form
  data-testid="addUserForm"
  onSubmit={[Function]}
>
  <input
    data-testid="name"
    onChange={[Function]}
    placeholder="Name"
    type="text"
    value=""
  />
  <p
    className="alert"
  />
  <input
    data-testid="email"
    onChange={[Function]}
    placeholder="Email"
    type="text"
    value=""
  />
  <p
    className="alert"
  />
  <input
    data-testid="number"
    onChange={[Function]}
    placeholder="Mobile"
    type="text"
    value=""
  />
  <p
    className="alert"
  />
  <div>
    <input
      checked={true}
      data-testid="checked"
      onChange={[Function]}
      onClick={[Function]}
      type="checkbox"
    />
    <p
      className="agree"
      data-testid="promotionsP"
    >
      Accept terms
    </p>
    <p
      className="alert"
    />
  </div>
  <button
    data-testid="submitButton"
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`<Form /> submits fails when validation errors  1`] = `
Array [
  <p
    className="alert"
  >
    Name is required
  </p>,
  <p
    className="alert"
  >
    Email is required
  </p>,
  <p
    className="alert"
  >
    Number is required
  </p>,
  <p
    className="alert"
  />,
]
`;

exports[`<Form /> submits the form 1`] = `"New User Added"`;
